### PR TITLE
New scrollbars for the scrollable refactor

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

File: packages/flutter/lib/src/material/scrollbar.dart [hunk 1]
```diff
@@ -2,10 +2,186 @@
...
 import 'package:flutter/widgets.dart';
 
 import 'theme.dart';
 
+class Scrollbar2 extends StatefulWidget {
+  Scrollbar2({
+    Key key,
+    this.child,
+  }) : super(key: key);
+
```
Comment (abarth) [inline packages/flutter/lib/src/material/scrollbar.dart]: Extra blank line.

File: packages/flutter/lib/src/material/scrollbar.dart [hunk 2]
```diff
@@ -2,10 +2,186 @@
...
+  void dispose() {
+    _controller.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return new NotificationListener<ScrollNotification2>(
+      onNotification: _handleScrollNotification,
+      child: new RepaintBoundary(
```
Comment (abarth) [inline packages/flutter/lib/src/material/scrollbar.dart]: I wonder if we should try to collapse out these repaint boundaries when the scroll bars are invisible.

Comment (abarth) [inline packages/flutter/lib/src/material/scrollbar.dart]: Maybe add a TODO about that?

File: packages/flutter/lib/src/material/scrollbar.dart [hunk 3]
```diff
@@ -2,10 +2,186 @@
...
+    super.dispose();
+  }
+
+  ScrollableMetrics _lastMetrics;
+  AxisDirection _lastAxisDirection;
+
+  static const double _kMinThumbExtent = 18.0;
+  static const double _kThumbGirth = 6.0;
+  static const Duration _kThumbFadeDuration = const Duration(milliseconds: 300);
+  static const Duration _kFadeOutTimeout = const Duration(milliseconds: 600);
```
Comment (abarth) [inline packages/flutter/lib/src/material/scrollbar.dart]: For private names like these, I've just been putting them as file-level private constants.

File: packages/flutter/lib/src/material/scrollbar.dart [hunk 4]
```diff
@@ -2,10 +2,186 @@
...
+        _paintThumb(_lastMetrics.extentBefore, _lastMetrics.extentInside, _lastMetrics.extentAfter, size.width, canvas, size, _paintHorizontalThumb);
+        break;
+      case AxisDirection.left:
+        _paintThumb(_lastMetrics.extentAfter, _lastMetrics.extentInside, _lastMetrics.extentBefore, size.width, canvas, size, _paintHorizontalThumb);
+        break;
+    }
+  }
+}
+
+class _ScrollbarPainter extends CustomPainter {
```
Comment (abarth) [inline packages/flutter/lib/src/material/scrollbar.dart]: Please file a bug about smoothing this path.

**Comment by abarth** [comment]
LGTM

File: packages/flutter/lib/src/material/scrollbar.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/scrollbar.dart]: is there some way we could make CustomPaint take a flag that decides whether to do it or not?

File: packages/flutter/lib/src/material/scrollbar.dart (hunk 3 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/scrollbar.dart]: I left them here so they're closer to the code using them. Doesn't seem to matter either way.

File: packages/flutter/lib/src/material/scrollbar.dart (hunk 4 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/scrollbar.dart]: https://github.com/flutter/flutter/issues/7648

**Pull request merged**